File: bin/import_db.js
#!/usr/bin/env node

'usr strict';

var microtime = function () {
  return new Date().valueOf()/1000; //: new Date().toISOString().replace(/[-T:Z]/g,'');
}

var _proc_mem = process.memoryUsage();
var _start_ts = microtime();

var glob     = require("glob"),  // github.com/isaacs/node-glob.git
    readline = require('readline'),
    fs       = require('fs'), 
    util     = require('util'), 
    url      = require('url'), 
    uap      = require('node-uap'),
    mysql    = require('mysql');

var connection = mysql.createConnection({
  socketPath : '/var/run/mysqld/mysqld.sock',
  host       : 'localhost',
  user       : 'root',
  password   : '',
  database   : ''
  //,debug    : true//;['ComQueryPacket', 'RowDataPacket']
});

connection.connect(function(err){
  if(err) {
    console.log("Error connecting database ...");    
    console.error(err);
  } else {
    console.log('mysql connected as id ' + connection.threadId);
    glob_file();
  }
});
connection.query('SET sql_log_bin = 0; /*LOCK TABLE access_logs WRITE;*/');

var i=0, rd_close=false;
var glob_file = function () { //{{{
  // search and list files
  glob("../../access/183.110.11.212*2017010500*-access_log", {}, function (err, files) {
    files.forEach( function (file) {
      console.log(file);
      var basename = file.substr( 1+file.lastIndexOf('/'));
      var host = basename.substr(0, basename.indexOf('-') );
      var rs = fs.createReadStream(file);
      rs.on('end', function () {
        rd_close = true;
        console.log('All the data in the file has been read');
        if ( global.gc ) {
          global.gc();
          console.log('garbage collected');
        }
      })
      var rd = readline.createInterface({
        input: rs/*,
        output: process.stdout,
        terminal: false,
        historySize: 0*/
      });

      rd.on('line', function(line) {
        ++i;
        //console.log('i:'+i);
        //if ( 0 == (i%5000) ) {
        //  console.log( "rd_line : "+i);
        //  console.log( "process.memoryUsage().rss : " + new Intl.NumberFormat().format(process.memoryUsage().rss - _proc_mem.rss) );
        //  console.log( "process.memoryUsage().headUsed : " + new Intl.NumberFormat().format(process.memoryUsage().heapUsed - _proc_mem.heapUsed) );
        //}
        //console.log(line);
        analyze_line(line,host);
      });
    });
  });
} //}}}

// analyze
var sql = 'INSERT INTO access_logs_myisam ( \
	host, ip, dt, method, req, protocol, code, byte, ref, ua, req_dir, req_base, req_query, req_frag, ref_host, ref_path, ref_query, ua_fam_maj, ua_full, os_fam_maj, os_full, dev_full \
) VALUES ( \
		 ?,  ?,  ?,      ?,   ?,        ?,    ?,    ?,   ?,  ?,       ?,        ?,         ?,        ?,        ?,        ?,         ?,          ?,       ?,          ?,       ?,        ? \
)';
var re = /([(\d\.)]+) - - \[(.*?)\] "([^\s]*?) ([^\s]*?) ([^\s]*?)" (\d+) (-|\d+) "(-|.*?)" "(.*?)"/;
var j=0, k=0;
var analyze_line = function (line,host) { //{{{
  var found = line.match(re);
  if ( found!==null && 10 <= found.length ) {
    var param = found.slice(0);
    param[0] = host;
    //param[1] = remote   = found[1];
    var d = found[2].substr(0, found[2].indexOf(':'));
    var t = found[2].substring( 1+found[2].indexOf(':'), found[2].indexOf('+') );
    param[2] = dt = new Date(Date.parse(d+' '+t+'UTC')).toISOString().substr(0,19).replace('T',' ');
    //param[3] = method   = found[3];
    //param[4] = req      = found[4];
    //param[5] = protocol = found[5];
    //param[6] = code     = found[6];
    //param[7] = byte     = found[7];
    //param[8] = ref      = found[8];
    //param[9] = ua       = found[9];

    req = url.parse(found[4]);
    param[10] = req.pathname.substr(0,req.pathname.lastIndexOf('/'));
    param[11] = req.pathname.substr(1+req.pathname.lastIndexOf('/')); //.split(/[\/]/).pop();
    param[12] = req.search; // default : null
    param[13] = req.hash;   // default : null

    ref = url.parse(found[4]);
    param[14] = ref.host;
    param[15] = ref.pathname; // default : null
    param[16] = ref.search;   // default : null

    // Parse everything
    var objUA = uap.parse(found[9]);
    param[17] = objUA.ua.family + (null==objUA.ua.major ? '' : ' '+objUA.ua.major);
    param[18] = param[17] + (null==objUA.ua.minor ? '' : '.'+objUA.ua.minor) +(null==objUA.ua.patch ? '' : '.'+objUA.ua.patch);
    param[19] = objUA.os.family + (null==objUA.os.major ? '' : ' '+objUA.os.major);
    param[20] = param[19] + (null==objUA.os.minor ? '' : '.'+objUA.os.minor) +(null==objUA.os.patch ? '' : '.'+objUA.os.patch);
    param[21] = objUA.device.family + (null==objUA.device.brand ? '' : ' '+objUA.device.brand) +(null==objUA.device.model ? '' : ' '+objUA.device.model);

    connection.query(sql,param, function(err,result) {
      ++j;
      //console.log('j:'+j);
      //if ( 0 == (j%5000) ) {
      //  console.log( "query: "+j+", skip: "+k );
      //  console.log( "process.memoryUsage().rss : " + new Intl.NumberFormat().format(process.memoryUsage().rss - _proc_mem.rss) );
      //  console.log( "process.memoryUsage().headUsed : " + new Intl.NumberFormat().format(process.memoryUsage().heapUsed - _proc_mem.heapUsed) );
      //}

      if ( err ) {
        console.log(line);
        console.log( 'FAILED : '+util.inspect(err) + "\n\tresults : " + util.inspect(result)+ "\n\tparam : " + util.inspect(param) );
        process.exit();
      } else if ( rd_close && i == (j+k) ) {
        console.log('connection.end');
        connection.end();
      }
    });
  } else {
    ++k;
    console.log('k:'+k);
    console.log(line);
    console.log('failed to parse req_url : '+line);
  }
} //}}}

File: bin/import_db.py
#!/usr/bin/python
import glob, os, mysql.connector, re, datetime
from mysql.connector import errorcode
from urlparse import urlparse
from ua_parser import user_agent_parser

config = {
  'unix_socket' : '/var/run/mysqld/mysqld.sock',
  'host': '127.0.0.1',
  'user': 'root',
  'password': '',
  'database': '',
  'autocommit' : True,
  'get_warnings': True,
  'raise_on_warnings': True,
  #'use_pure': False,
}

try:
  cnx = mysql.connector.connect(**config)
except mysql.connector.Error as err:
  if err.errno == errorcode.ER_ACCESS_DENIED_ERROR:
    print("Something is wrong with your user name or password")
  elif err.errno == errorcode.ER_BAD_DB_ERROR:
    print("Database does not exist")
  else:
    print(err)
    exit(1)
#else:
#  print('nothing')
#  cnx.close()
#  exit(1)

cursor = cnx.cursor() #prepared=True
cursor.execute('SET sql_log_bin = 0; /*LOCK TABLE access_logs WRITE;*/', multi=True)

sql = ("INSERT INTO access_logs_myisam ("
    "host, ip, dt, method, req, protocol, code, byte, ref, ua, req_dir, req_base, req_query, req_frag, ref_host, ref_path, ref_query, ua_fam_maj, ua_full, os_fam_maj, os_full, dev_full"
    ") VALUES ("
    "  %s, %s, %s,     %s,  %s,       %s,   %s,   %s,  %s, %s,      %s,       %s,        %s,       %s,       %s,       %s,        %s,         %s,      %s,         %s,      %s,       %s"
    ")")

pat = re.compile( '([(\d\.)]+) - - \[(.*?)\] "([^\s]*?) ([^\s]*?) ([^\s]*?)" (\d+) (-|\d+) "(-|.*?)" "(.*?)' )

def analyze (line, host):
  global pat, sql, cursor
  found = pat.findall(line)
  #print found
  if not found :
    print ('failed to parse req_url : ' + line);
    return
  elif len(found[0]) < 9 :
    print ('insufficient req_url : ' + line);
    print found[0]
    return

  (remote,dt_old,method,req,protocol,code,byte,ref,ua) = found[0]
  dt = datetime.datetime.strptime(dt_old[:dt_old.find(' ')],'%d/%b/%Y:%H:%M:%S')
  dt_new = dt.__str__()
  if byte == '-' :
    byte = None
  #print (remote,dt_new,method,req,protocol,code,byte,ref,ua)

  oReq = urlparse(req)
  req_dir, req_base = os.path.split(oReq.path)
  req_query = oReq.query
  req_frag = oReq.fragment
  #print (req_dir, req_base,req_query,req_frag)

  oRef = urlparse(ref)
  ref_host = oRef.netloc
  ref_path = oRef.path
  ref_query = oRef.query # oRef.params
  #print (ref_host, ref_path,ref_query)

  if ''==ua.strip() :
    ua_fam_maj = ua_full = os_fam_maj = os_full = dev_full = None
  else :
    oUA = user_agent_parser.Parse(ua)
    ua_fam_maj = oUA['user_agent']['family']or'' + ' ' +oUA['user_agent']['major']or'' 
    ua_full    = ( oUA['user_agent']['family']or'' + ' ' +oUA['user_agent']['major']or'' +'.'+ oUA['user_agent']['minor']or'' +'.'+ oUA['user_agent']['patch']or'' ).strip('.')
    os_fam_maj = oUA['os']['family']or'' + ' ' +oUA['os.major']
    os_full    = ( oUA['os']['family']or'' + ' ' +oUA['os.major']or''  +'.'+ oUA['os.minor']or'' +'.'+ oUA['os.patch']or'' ).strip('.')
    dev_full   = ( oUA['device']['family']or'' + ' ' +oUA['device']['brand']or'' +' '+ oUA['device']['model']or'' ).strip()
  #print ua,':',ua_fam_maj,ua_full,os_fam_maj,os_full,dev_full

  param = ( host, remote, dt, method, req, protocol, code, byte, ref, ua, req_dir, req_base, req_query, req_frag, ref_host, ref_path, ref_query, ua_fam_maj, ua_full, os_fam_maj, os_full, dev_full )
  #print param

  cursor.execute(sql, param )

  return

file_pattern = '../../access/183.110.11.212*2017010500*-access_log'
for onefile in glob.glob(file_pattern) :
  basename = onefile[ 1+onefile.rfind('/'): ]
  host = basename[: basename.find('-') ]
  #print (onefile,basename,host)
  with open(onefile, 'r+') as f:
    i = 0
    for line in f:
      #print line
      analyze (line,host)
      i+=1
      if 0 == (i % 5000) :
        print i

cursor.close()
cnx.close()

